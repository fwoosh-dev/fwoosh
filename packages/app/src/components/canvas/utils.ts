import potpack from "potpack";
import { TLBounds, Utils } from "@tldraw/core";
import Vec from "@tldraw/vec";
import { StorySidebarChildItem } from "@fwoosh/types";

import { GroupShape } from "./shapes/group";
import { Shape } from "./shapes";
import { group } from "console";

export function createGroup(
  props: Pick<GroupShape, "id" | "childIndex">
): GroupShape {
  return {
    ...props,
    name: props.id,
    type: "group",
    parentId: "canvas",
    rotation: 0,
    childIds: [],
    stories: [],
    size: [0, 0],
    contentSize: [0, 0],
    hasBeenMeasured: false,
    point: [0, 0],
  };
}

export function getGroupBounds(shape: GroupShape): TLBounds {
  const [width, height] = shape.size;

  const bounds = {
    minX: 0,
    maxX: width,
    minY: 0,
    maxY: height,
    width,
    height,
  };

  return Utils.translateBounds(bounds, shape.point);
}

export const getBounds = {
  group: getGroupBounds,
  docs: getGroupBounds,
};

function convertShapesToPotpackData(data: Shape[]) {
  return data.map((i) => {
    return {
      h: i.size[1] + 16,
      w: i.size[0] + 16,
      x: i.point[0],
      y: i.point[1],
      id: i.id,
    };
  });
}

/** Recursively move all the shapes in a group. */
function updateChildPositions(
  group: GroupShape,
  shapes: Record<string, Shape>,
  moveDistance: number[]
) {
  group.childIds.forEach((child) => {
    const childShape = shapes[child];

    if (!childShape) {
      return;
    }

    childShape.point = Vec.add(childShape.point, moveDistance);

    /// BUG it doesn't seem to update the nested group position
    if (childShape.type === "group") {
      updateChildPositions(childShape, shapes, moveDistance);
    }
  });
}

/**
 * Pack the groups into a square shape and update the position of the
 * children to be relative to the group.
 *
 * Optionally pass an offset to the group position to account for
 * the groups itself having a content size.
 */
function packGroup(
  boxes: GroupShape[],
  shapes: Record<string, Shape>,
  offset: number[] = [0, 0]
) {
  const data = convertShapesToPotpackData(boxes);

  potpack(data);

  data.forEach((i) => {
    const shape = shapes[i.id];
    const oldPoint = shape.point;

    // Make the point below the groups content if it has it
    shape.point = Vec.add([i.x, i.y], offset);
    const moveDistance = Vec.sub(shape.point, oldPoint);

    if (shape.type === "group") {
      updateChildPositions(shape, shapes, moveDistance);
    }
  });

  return data;
}

/**
 * Iterates over the tree and take items with type="tree" and
 * packs them into groups.
 *
 * It will then update the size of the group to fit the children
 * so that the group can be used as a container for the children.
 * This container is then used to pack those groups into bigger and
 * bigger groups.
 */
export function packShapesIntoGroups(
  items: StorySidebarChildItem[],
  shapes: Record<string, Shape>
) {
  const boxes: GroupShape[] = [];

  for (const item of items) {
    const shape = shapes[item.id];

    if (!shape || shape.type === "docs" || item.type === "story") {
      // if (item.story.type === "basic") {
      //   // boxes.push(shapes[item.id]);
      // }
      continue;
    } else {
      const childrenBoxes = packShapesIntoGroups(item.children, shapes);
      const data = packGroup(
        [
          ...(shape.childIds.length > 0
            ? [{ ...shape, size: shape.contentSize }]
            : []),
          ...childrenBoxes,
        ],
        shapes,
        [0, shape.contentSize[1] + 16]
      );

      const childBounds = data.map((i) => {
        const shape = shapes[i.id];
        return getBounds[shape.type](shape as any);
      });
      const bounds = Utils.getCommonBounds(childBounds);

      if (bounds) {
        shape.size = [bounds.width, bounds.height];
      }

      boxes.push(shape);
    }
  }

  packGroup(boxes, shapes);

  return boxes;
}
