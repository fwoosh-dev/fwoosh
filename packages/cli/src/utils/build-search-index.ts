import fs from "fs";
import { relative, join } from "path";
import * as cheerio from "cheerio";
import { Element } from "domhandler";

import glob from "fast-glob";

const getContentBeforeNextHeading = (
  $: ReturnType<typeof cheerio.load>,
  node: cheerio.Element
) => {
  const next = $(node).nextAll().toArray();
  let content = "";

  for (let index = 0; index < next.length; index++) {
    const element = next[index];

    if ($(element).attr("class")?.includes("lvl")) {
      break;
    }

    content += `\n${$(element).text()}`;
  }

  return content;
};

const getHeadingsBeforeNextHeading = (
  $: ReturnType<typeof cheerio.load>,
  node: cheerio.Element,
  lvl: number
) => {
  const next = $(node).nextAll().toArray();
  const titles = [];

  for (let index = 0; index < next.length; index++) {
    const element = next[index];
    const $el = $(element);

    if ($el.hasClass(`lvl${lvl}`)) {
      break;
    }

    if ($el.hasClass(`lvl${lvl + 1}`)) {
      titles.push(element);
    }
  }

  return titles;
};

// const getPage = (page: string) => {
//   const content = fs.readFileSync(page, "utf-8");
//   const $ = cheerio.load(content);

//   const lvl0Node = $(".lvl0")
//     .toArray()
//     .find((title) => $(".sidebar-active", $(title.next)).length);
//   const isBlog = page.includes("/blog/");
//   const lvl0 = isBlog ? "Blog" : $(lvl0Node).text();

//   const buildTree = (parents: string[], node: Element, lvl = 1) => {
//     const searchNode = isBlog && lvl === 1 ? node.parent : node;
//     const nextHeadings = getHeadingsBeforeNextHeading($, searchNode, lvl);
//     const path = [...parents, $(node).text()];

//     const currentNode = {
//       path,
//       url: `${relative(DOCS_DIR, page)
//         .replace("out/", "")
//         .replace(".html", "")}${node.attribs.id ? `#${node.attribs.id}` : ""}`,
//       content: getContentBeforeNextHeading($, searchNode),
//     };

//     return [
//       currentNode,
//       ...nextHeadings.reduce(
//         (acc, heading) => [...acc, ...buildTree(path, heading, lvl + 1)],
//         []
//       ),
//     ];
//   };

//   const firstLvl = [1, 2, 3, 4, 5, 6]
//     .map((lvl) => $(`.lvl${lvl}`)[0])
//     .find(Boolean);

//   if (!firstLvl) {
//     return;
//   }

//   return buildTree([lvl0], firstLvl);
// };

// export const buildSearchIndex = (dest = "out") => {
//   return glob(join(DOCS_DIR, "out/**/*.html")).then((pages) => {
//     const pagesToIndex = pages.filter(
//       (page) =>
//         !page.includes("/_sidebar") &&
//         !page.includes("out/blog.html") &&
//         !page.includes("out/404.html") &&
//         !page.includes("out/index.html")
//     );

//     fs.writeFileSync(
//       join(DOCS_DIR, dest, "search-index.json"),
//       JSON.stringify(pagesToIndex.map(getPage).flat().filter(Boolean), null, 2)
//     );
//   });
// };
