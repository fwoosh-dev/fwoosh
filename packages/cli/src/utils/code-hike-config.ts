import { createRequire } from "module";
import fs from "fs";
import { remarkCodeHike } from "@code-hike/mdx";
import { SyntaxTheme } from "@fwoosh/types";

let syntaxTheme: SyntaxTheme = "nord";

export function setSyntaxTheme(theme: SyntaxTheme) {
  syntaxTheme = theme;
}

const require = createRequire(import.meta.url);

export const getCodeHikeConfig = () => {
  const themeObject = JSON.parse(
    fs.readFileSync(require.resolve(`shiki/themes/${syntaxTheme}.json`), "utf8")
  );

  return {
    theme: themeObject,
    showCopyButton: true,
    skipLanguages: [],
  } as Parameters<typeof remarkCodeHike>[0];
};
