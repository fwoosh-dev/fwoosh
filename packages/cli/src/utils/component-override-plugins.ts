interface ComponentOverridePluginOptions {
  componentOverrides?: string;
}

export const componentOverridePlugin = (
  options: ComponentOverridePluginOptions
) => {
  const virtualFileId = "@fwoosh/app/overrides";

  return {
    name: "component-overrides",

    resolveId(id: string) {
      if (id.includes(virtualFileId)) {
        return virtualFileId;
      }

      return null;
    },

    async load(id: string) {
      if (id.includes(virtualFileId)) {
        if (options.componentOverrides) {
          return `export * from "${options.componentOverrides}"`;
        }

        return "";
      }

      return;
    },
  };
};
