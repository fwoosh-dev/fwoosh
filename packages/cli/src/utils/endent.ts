import endentPackage from "endent";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export const endent = (endentPackage as any).default as typeof endentPackage;
