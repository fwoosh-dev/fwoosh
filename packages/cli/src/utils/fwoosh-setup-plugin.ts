/** Plugin that injects fwoosh config options into the front-end */
export function fwooshSetupPlugin({ file }: { file: string }) {
  const virtualFileId = "@fwoosh/app/setup";

  return {
    name: "fwoosh-setup",

    resolveId(id: string) {
      if (id.includes(virtualFileId)) {
        return virtualFileId;
      }

      return null;
    },

    async load(id: string) {
      if (id.includes(virtualFileId)) {
        return file ? `import "${file}"` : "";
      }
      return;
    },
  };
}
