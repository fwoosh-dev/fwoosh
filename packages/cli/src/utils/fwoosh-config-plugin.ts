import { FwooshOptionsLoaded } from "@fwoosh/types";
import { endent } from "./endent.js";

/** Plugin that injects fwoosh config options into the front-end */
export function fwooshConfigPlugin(config: FwooshOptionsLoaded) {
  const virtualFileId = "@fwoosh/app/config";

  return {
    name: "fwoosh-config",

    resolveId(id: string) {
      if (id.includes(virtualFileId)) {
        return virtualFileId;
      }

      return null;
    },

    async load(id: string) {
      if (id.includes(virtualFileId)) {
        return endent`
          export const config = {
            title: "${config.title}",
            includeMdxInWorkbench: ${config.includeMdxInWorkbench},
            themes: ${JSON.stringify(config.themes)},
          }
        `;
      }
      return;
    },
  };
}
