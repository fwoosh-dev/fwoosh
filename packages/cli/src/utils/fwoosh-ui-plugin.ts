import { endent } from "./endent.js";
import { pascalCase } from "change-case";
import { FwooshPanel } from "@fwoosh/types";

function lazyLoadComponents(
  imports: {
    name?: string;
    filepath: string;
    hideWithoutParams?: string | false;
  }[]
) {
  const names: string[] = [];
  const components = imports.map((i) => {
    const name = i.name || pascalCase(i.filepath);
    names.push(name);

    return endent`
      const ${name} = React.lazy(() => import('${i.filepath}').then(m => {
        if (m.Name) {
          m.default.panelTitle = m.Name;
        }
        return m;
      }));
      ${name}.componentName = '${name}';
      ${name}.displayName = React.lazy(() => import('${i.filepath}').then(m => {
        return { default: m.Name || (() => '${name}') };
      }));
      ${
        i.hideWithoutParams
          ? `${name}.hideWithoutParams = "${i.hideWithoutParams}";`
          : ""
      }
    `;
  });

  return { names, components };
}

/** Plugin that injects fwoosh config options into the front-end */
export function fwooshUiPlugin({
  panels,
  toolbarControls,
}: {
  panels: FwooshPanel[];
  toolbarControls: string[];
}) {
  const virtualFileId = "@fwoosh/app/ui";

  return {
    name: "fwoosh-config",

    resolveId(id: string) {
      if (id.includes(virtualFileId)) {
        return virtualFileId;
      }

      return null;
    },

    async load(id: string) {
      if (id.includes(virtualFileId)) {
        const toolbar = lazyLoadComponents(
          toolbarControls.map((filepath) => ({ filepath }))
        );
        const panel = lazyLoadComponents(panels);

        return endent`
          import * as React from "react";

          ${toolbar.components.join("")}
          export const toolbarControls = [ ${toolbar.names.join(", ")}];

          ${panel.components.join("")}
          export const panels = [ ${panel.names.join(", ")}];
        `;
      }
      return;
    },
  };
}
