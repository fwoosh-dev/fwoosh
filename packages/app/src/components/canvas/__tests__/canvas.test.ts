import { expect, describe, test } from "vitest";

import twoTree from "./data/two-tree.json";

import { createGroup, packShapesIntoGroups } from "../utils";

describe("packShapesIntoGroups", () => {
  test("it should group two shapes", () => {
    const group1 = createGroup({ id: "1", childIndex: 1 });
    group1.size = [100, 100];

    const group2 = createGroup({ id: "2", childIndex: 2 });
    group2.size = [100, 100];

    const shapes = {
      1: group1,
      2: group2,
    };

    packShapesIntoGroups(
      [
        { id: "1", children: [], name: "one", type: "tree" },
        { id: "2", children: [], name: "two", type: "tree" },
      ],
      {
        1: group1,
        2: group2,
      }
    );

    expect(shapes).toMatchSnapshot();
  });

  test("it should wrap single item in nested group", () => {
    const group1 = createGroup({ id: "1", childIndex: 1 });
    group1.size = [100, 100];

    const group2 = createGroup({ id: "2", childIndex: 2 });
    group2.size = [100, 100];

    group1.children = [group2.id];

    const shapes = {
      1: group1,
      2: group2,
    };

    packShapesIntoGroups(
      [
        {
          id: "1",
          children: [{ id: "2", children: [], name: "two", type: "tree" }],
          name: "one",
          type: "tree",
        },
      ],
      shapes
    );

    expect(shapes).toMatchSnapshot();
  });

  test("it should wrap single item in nested group with content size", () => {
    const group1 = createGroup({ id: "1", childIndex: 1 });
    group1.size = [100, 100];
    group1.contentSize = [100, 100];

    const group2 = createGroup({ id: "2", childIndex: 2 });
    group2.size = [100, 100];

    group1.children = [group2.id];

    const shapes = {
      1: group1,
      2: group2,
    };

    packShapesIntoGroups(
      [
        {
          id: "1",
          children: [{ id: "2", children: [], name: "two", type: "tree" }],
          name: "one",
          type: "tree",
        },
      ],
      shapes
    );

    expect(shapes).toMatchSnapshot();
  });

  test("it should be correct for a group with two children beside each other", () => {
    const group1 = createGroup({ id: "1", childIndex: 1 });
    group1.size = [100, 100];
    group1.contentSize = [100, 100];

    const group2 = createGroup({ id: "2", childIndex: 2 });
    group2.size = [100, 100];

    const group3 = createGroup({ id: "3", childIndex: 2 });
    group3.size = [100, 100];

    group1.children = [group2.id, group3.id];

    const shapes = {
      1: group1,
      2: group2,
      3: group3,
    };

    packShapesIntoGroups(
      [
        {
          id: "1",
          children: [
            { id: "2", children: [], name: "two", type: "tree" },
            { id: "3", children: [], name: "three", type: "tree" },
          ],
          name: "one",
          type: "tree",
        },
      ],
      shapes
    );

    expect(shapes).toMatchSnapshot();
  });

  test("it should work with 2 levels of nesting", () => {
    const group1 = createGroup({ id: "1", childIndex: 1 });
    group1.size = [100, 100];
    group1.contentSize = [100, 100];

    const group2 = createGroup({ id: "2", childIndex: 2 });
    group2.size = [100, 100];
    group2.contentSize = [100, 100];

    const group3 = createGroup({ id: "3", childIndex: 2 });
    group3.size = [100, 100];

    group2.children = [group3.id];
    group1.children = [group2.id];

    const shapes = {
      1: group1,
      2: group2,
      3: group3,
    };

    packShapesIntoGroups(
      [
        {
          id: "1",
          children: [
            {
              id: "2",
              children: [
                { id: "3", children: [], name: "three", type: "tree" },
              ],
              name: "two",
              type: "tree",
            },
          ],
          name: "one",
          type: "tree",
        },
      ],
      shapes
    );

    expect(shapes).toMatchSnapshot();
  });

  test("it should work with 2 groups of 2 levels of nesting", () => {
    const group1 = createGroup({ id: "1", childIndex: 1 });
    group1.size = [100, 100];
    group1.contentSize = [100, 100];

    const group2 = createGroup({ id: "2", childIndex: 2 });
    group2.size = [100, 100];
    group2.contentSize = [100, 100];

    const group3 = createGroup({ id: "3", childIndex: 3 });
    group3.size = [100, 100];

    const group4 = createGroup({ id: "4", childIndex: 4 });
    group4.size = [100, 100];
    group4.contentSize = [100, 100];

    const group5 = createGroup({ id: "5", childIndex: 5 });
    group5.size = [100, 100];

    group2.children = [group3.id];
    group4.children = [group5.id];
    group1.children = [group2.id, group4.id];

    const shapes = {
      1: group1,
      2: group2,
      3: group3,
      4: group4,
      5: group5,
    };

    packShapesIntoGroups(
      [
        {
          id: "1",
          children: [
            {
              id: "2",
              children: [
                { id: "3", children: [], name: "three", type: "tree" },
              ],
              name: "two",
              type: "tree",
            },
            {
              id: "4",
              children: [{ id: "5", children: [], name: "five", type: "tree" }],
              name: "four",
              type: "tree",
            },
          ],
          name: "one",
          type: "tree",
        },
      ],
      shapes
    );

    expect(shapes).toMatchSnapshot();
  });

  test("it should work long and wide children", () => {
    packShapesIntoGroups(twoTree.tree as any, twoTree.shapes as any);
    expect(twoTree.shapes).toMatchSnapshot();
  });
});
