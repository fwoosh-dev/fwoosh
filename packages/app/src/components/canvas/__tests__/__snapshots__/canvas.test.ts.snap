// Vitest Snapshot v1

exports[`packShapesIntoGroups > it should work long and wide children 1`] = `
{
  "Components": {
    "childIndex": 0,
    "children": [
      "Components-Button",
      "Components-Text",
    ],
    "contentSize": [
      148.421875,
      70,
    ],
    "hasBeenMeasured": true,
    "id": "Components",
    "name": "Components",
    "parentId": "canvas",
    "point": [
      0,
      0,
    ],
    "rotation": 0,
    "size": [
      633.4921875,
      1059,
    ],
    "stories": [],
    "type": "group",
  },
  "Components-Button": {
    "childIndex": 0,
    "children": [
      "Components-Button-Button",
    ],
    "contentSize": [
      89.7265625,
      70,
    ],
    "hasBeenMeasured": true,
    "id": "Components-Button",
    "name": "Components-Button",
    "parentId": "canvas",
    "point": [
      366.015625,
      86,
    ],
    "rotation": 0,
    "size": [
      267.4765625,
      516,
    ],
    "stories": [],
    "type": "group",
  },
  "Components-Button-Button": {
    "childIndex": 0,
    "children": [],
    "contentSize": [
      267.4765625,
      516,
    ],
    "hasBeenMeasured": true,
    "id": "Components-Button-Button",
    "name": "Components-Button-Button",
    "parentId": "canvas",
    "point": [
      0,
      86,
    ],
    "rotation": 0,
    "size": [
      267.4765625,
      516,
    ],
    "stories": [
      "components-button-button--playground",
      "components-button-button--with-disabled",
      "components-button-button--with-style",
      "components-button-button--without-style",
    ],
    "type": "group",
  },
  "Components-Text": {
    "childIndex": 1,
    "children": [],
    "contentSize": [
      350.015625,
      1059,
    ],
    "hasBeenMeasured": true,
    "id": "Components-Text",
    "name": "Components-Text",
    "parentId": "canvas",
    "point": [
      0,
      86,
    ],
    "rotation": 0,
    "size": [
      350.015625,
      1059,
    ],
    "stories": [
      "components-text--headings",
      "components-text--unordered-list",
      "components-text--ordered-list",
      "components-text--anchor",
    ],
    "type": "group",
  },
}
`;
